File: leetcode/medium/ContainerWithMostWater.py
from typing import List

class Solution:
    def maxArea(self, heights: List[int]) -> int:
        """
        Finds two lines that together with the x-axis form a container,
        such that the container contains the most water.

        The problem is solved using the "two-pointer" approach.
        We start with two pointers, one at the beginning and one at the end of the array.
        The area formed by these two lines is calculated.
        Then, we move the pointer pointing to the shorter line inwards.
        This is because moving the taller line inwards will never increase the height
        (it might decrease or stay the same), but it will definitely decrease the width.
        Moving the shorter line inwards might find a taller line, potentially increasing the area.

        Args:
            heights: A list of non-negative integers representing the heights of the vertical lines.

        Returns:
            The maximum amount of water that can be contained.
        """
        
        # Initialize two pointers: 'left' at the beginning and 'right' at the end of the array.
        left, right = 0, len(heights) - 1
        
        # Initialize 'max_area' to store the maximum area found so far.
        max_area = 0
        
        # Continue as long as the left pointer is less than the right pointer.
        while left < right:
            # Calculate the current area: width * height.
            # Width is the distance between the two pointers (right - left).
            # Height is limited by the shorter of the two lines (min(heights[left], heights[right])).
            current_area = (right - left) * min(heights[left], heights[right])
            
            # Update max_area if the current_area is greater.
            max_area = max(max_area, current_area)
            
            # Move the pointer that points to the shorter line inwards.
            # This strategy is crucial because to potentially increase the area,
            # we need to find a taller line. Moving the shorter line gives us
            # a chance to find a taller line, while moving the taller line
            # would only decrease the width without guarantee of increasing height.
            if heights[left] > heights[right]:
                right -= 1
            else:
                left += 1
                
        return max_area

File: leetcode/medium/ContainerWithMostWater.java
class Solution {

    /**
     * Calculates the maximum amount of water a container can store.
     * * <p>Algorithm: Two-Pointer Approach
     * <br>Time Complexity: O(n) - We traverse the array once.
     * <br>Space Complexity: O(1) - We use constant extra space.
     * * @param height An array of non-negative integers representing the height of lines.
     * @return The maximum area of water that can be contained.
     */
    public int maxArea(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int maxArea = 0;

        while (left < right) {
            // Cache values to avoid repeated array access
            int hLeft = height[left];
            int hRight = height[right];

            // Calculate area: width * min_height
            // We use the shorter line because water would spill over the taller one
            int currentArea = (right - left) * Math.min(hLeft, hRight);
            
            // Update maxArea if the current one is larger
            if (currentArea > maxArea) {
                maxArea = currentArea;
            }

            // Move the pointer of the shorter line inward.
            // Logic: Moving the taller pointer would decrease the width without 
            // potentially increasing the height (bounded by the shorter line),
            // guaranteeing a smaller area. We must move the shorter line to hope for a taller one.
            if (hLeft < hRight) {
                left++;
                // Optimization: Skip lines shorter than or equal to the current left line
                // because the width is shrinking. We need a strictly taller line to beat the current area.
                while (left < right && height[left] <= hLeft) {
                    left++;
                }
            } else {
                right--;
                // Optimization: Skip lines shorter than or equal to the current right line
                while (left < right && height[right] <= hRight) {
                    right--;
                }
            }
        }
        
        return maxArea;
    }
}